src/main/java/ganymedes01/etfuturum/blocks/BlockBarrel.java

```diff
@@ -58,26 +71,63 @@ public IIcon getIcon(int side, int meta) {
 
 	@Override
 	public void onBlockPlacedBy(World worldIn, int x, int y, int z, EntityLivingBase placer, ItemStack itemIn) {
+		TileEntityBarrel box = (TileEntityBarrel) worldIn.getTileEntity(x, y, z);
+		if (itemIn.hasTagCompound()) {
+			if (itemIn.hasDisplayName()) {
+				box.setCustomName(itemIn.getDisplayName()); // setCustomName
+			}
+		}
+
 		int l = BlockPistonBase.determineOrientation(worldIn, x, y, z, placer);
 		worldIn.setBlockMetadataWithNotify(x, y, z, l, 2);
 	}
 
 	@Override
 	@SideOnly(Side.CLIENT)
 	public void registerBlockIcons(IIconRegister i) {
-		this.blockIcon = i.registerIcon(getTextureName() + "_side");
-		this.topIcon = i.registerIcon(getTextureName() + "_top");
-		this.innerTopIcon = i.registerIcon(getTextureName() + "_top_open");
-		this.bottomIcon = i.registerIcon(getTextureName() + "_bottom");
+		if (type == TileEntityBarrel.BarrelType.VANILLA){
+			this.blockIcon = i.registerIcon(getTextureName() + "_side");
+			this.topIcon = i.registerIcon(getTextureName() + "_top");
+			this.innerTopIcon = i.registerIcon(getTextureName() + "_top_open");
+			this.bottomIcon = i.registerIcon(getTextureName() + "_bottom");
+		}else {
+			this.blockIcon = i.registerIcon(String.format("metalbarrels:%s_barrel_side", type.name().toLowerCase()));
+			this.topIcon = i.registerIcon(String.format("metalbarrels:%s_barrel_top", type.name().toLowerCase()));
+			this.innerTopIcon = i.registerIcon(String.format("metalbarrels:%s_barrel_top_open", type.name().toLowerCase()));
+			this.bottomIcon = i.registerIcon(String.format("metalbarrels:%s_barrel_bottom", type.name().toLowerCase()));
+		}
```

Comment:
See the comment on the constructor about building the texture name.
```suggestion
		this.blockIcon = i.registerIcon(getTextureName() + "_side");
		this.topIcon = i.registerIcon(getTextureName() + "_top");
		this.innerTopIcon = i.registerIcon(getTextureName() + "_top_open");
		this.bottomIcon = i.registerIcon(getTextureName() + "_bottom");
```